### PR TITLE
Add latent novelty query strategy using OODGuard

### PR DESCRIPTION
**Pull request by ktangsali**
# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

**Review by mnabian — APPROVED** [review]
Changes to the ODD guard look good to me.

File: examples/cfd/external_aerodynamics/active_learning_aero/src/run_al.py
Comment (laserkelvin) [inline examples/cfd/external_aerodynamics/active_learning_aero/src/run_al.py]: Stylistically, I would suggest writing this as a `match/case` statement instead of a bunch of `if/elif`

File: examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py [hunk 1]
```diff
@@ -368,6 +369,326 @@ def sample(self, query_queue: AbstractQueue, *args: Any, **kwargs: Any) -> None:
...
+
+        local_t = torch.cat(local_rows, dim=0)
+        gathered = padded_all_gather(local_t, device)
+        # Dedupe by flat_idx (column 0). This is a no-op for the
+        # round-robin shard mode (shards are disjoint) and removes the
+        # replicated rows produced by the n_total < world_size branch.
+        seen: set[int] = set()
+        keep_mask = torch.zeros(gathered.shape[0], dtype=torch.bool, device=device)
+        for i in range(gathered.shape[0]):
+            fid = int(gathered[i, 0].item())
```
Comment (laserkelvin) [inline examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py]: Not sure if you want to look at re-writing this so that it's not a blocking call with `item()`

File: examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py [hunk 2]
```diff
@@ -368,6 +369,326 @@ def sample(self, query_queue: AbstractQueue, *args: Any, **kwargs: Any) -> None:
...
+        backbone.eval()
+        embedding_reduction.eval()
+
+        # ---- Phase 1: calibrate OODGuard on labeled latents ----
+        if rank == 0:
+            self.logger.info(
+                f"Calibrating OODGuard on {len(labeled)} labeled samples..."
+            )
+        labeled_table = self._embed_indices(
+            [int(i) for i in labeled.tolist()],
```
Comment (laserkelvin) [inline examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py]: Why not keep this as a tensor, since the method accepts a tensor?

File: examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py [hunk 3]
```diff
@@ -368,6 +369,326 @@ def sample(self, query_queue: AbstractQueue, *args: Any, **kwargs: Any) -> None:
...
+            reduced = embedding_reduction(embedding_states.flatten(1, 2))
+            novelty = guard.score_geometry(reduced.detach().to(torch.float32))
+            local_rows.append([float(flat_idx), float(novelty.item())])
+
+        local_t = torch.tensor(local_rows, dtype=torch.float64, device=device)
+        if local_t.ndim == 1:
+            local_t = local_t.unsqueeze(0)
+        all_data = padded_all_gather(local_t, device).cpu().numpy()
+
+        scores = [(int(row[0]), float(row[1])) for row in all_data]
```
Comment (laserkelvin) [inline examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py:674]: Would be good to have a comment line here just to indicate what is being unpacked

File: physicsnemo/experimental/guardrails/embedded/ood_guard.py [hunk 4]
```diff
@@ -198,6 +198,67 @@ def check(
             self._threshold_stale = False
         self._check_geometry(geometry_latent)
 
+    @torch.compiler.disable
+    @torch.no_grad()
+    def score_geometry(self, geometry_latent: torch.Tensor) -> torch.Tensor:
```
Comment (laserkelvin) [inline physicsnemo/experimental/guardrails/embedded/ood_guard.py]: Add jaxtyping shape and dtype hints to input and output

**Review by laserkelvin — CHANGES_REQUESTED** [review]
Minor comments

File: physicsnemo/experimental/guardrails/embedded/ood_guard.py (hunk 4 above)
Comment (ktangsali) [reply physicsnemo/experimental/guardrails/embedded/ood_guard.py]: Thanks, addressed here: https://github.com/NVIDIA/physicsnemo/pull/1678/commits/3b32b04380a2450e61d46639be628cf09af5d371

File: examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py (hunk 3 above)
Comment (ktangsali) [reply examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py:674]: Thanks, addressed here: https://github.com/NVIDIA/physicsnemo/pull/1678/commits/3b32b04380a2450e61d46639be628cf09af5d371

File: examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py (hunk 2 above)
Comment (ktangsali) [reply examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py]: Done here: https://github.com/NVIDIA/physicsnemo/pull/1678/commits/3b32b04380a2450e61d46639be628cf09af5d371

File: examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py (hunk 1 above)
Comment (ktangsali) [reply examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py]: Done here: https://github.com/NVIDIA/physicsnemo/pull/1678/commits/3b32b04380a2450e61d46639be628cf09af5d371

File: examples/cfd/external_aerodynamics/active_learning_aero/src/run_al.py
Comment (ktangsali) [reply examples/cfd/external_aerodynamics/active_learning_aero/src/run_al.py]: Done: https://github.com/NVIDIA/physicsnemo/pull/1678/commits/3b32b04380a2450e61d46639be628cf09af5d371

File: physicsnemo/experimental/guardrails/embedded/ood_guard.py [hunk 5]
```diff
@@ -198,6 +198,67 @@ def check(
...
+                "in-distribution sample first."
+            )
+        # Upcast so cdist against the fp32 store works under AMP inputs.
+        pooled = geometry_latent.detach().to(self.geo_embeddings.dtype)
+        z = pooled / (pooled.norm(dim=-1, keepdim=True) + 1e-8)
+        store = self.geo_embeddings[:n_valid]
+        store_norm = store / (store.norm(dim=-1, keepdim=True) + 1e-8)
+        k = min(self.knn_k, n_valid)
+        _, dists = knn(store_norm, z, k)  # (B, k)
+        return dists.mean(dim=-1)  # (B,)
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply physicsnemo/experimental/guardrails/embedded/ood_guard.py:266]: Done here: https://github.com/NVIDIA/physicsnemo/commit/3b32b04380a2450e61d46639be628cf09af5d371

File: examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py [hunk 6]
```diff
@@ -368,6 +369,326 @@ def sample(self, query_queue: AbstractQueue, *args: Any, **kwargs: Any) -> None:
...
+                return_embedding_states=True,
+            )
+            reduced = embedding_reduction(embedding_states.flatten(1, 2))
+            novelty = guard.score_geometry(reduced.detach().to(torch.float32))
+            local_rows.append([float(flat_idx), float(novelty.item())])
+
+        local_t = torch.tensor(local_rows, dtype=torch.float64, device=device)
+        if local_t.ndim == 1:
+            local_t = local_t.unsqueeze(0)
+        all_data = padded_all_gather(local_t, device).cpu().numpy()
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply examples/cfd/external_aerodynamics/active_learning_aero/src/strategies.py]: Done here: https://github.com/NVIDIA/physicsnemo/commit/3b32b04380a2450e61d46639be628cf09af5d371

**Review by laserkelvin — APPROVED** [review]
LGTM

**Comment by ktangsali** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1678/changes/62f7fedf7d00495cf2486e7330c967117a492456
